Make it possible to switch between BC and BC-FIPS libraries

pedroigor opened the pull request:
Closes #12424

* After the last Quarkus upgrade, we are now able to remove unwanted dependencies from the runtime
* Introduces a new `security` configuration category to serve as an umbrella for security-related settings like FIPS.
* Introduces a new `security-fips-mode` option to set the FIPS mode. Allowed values are `enabled`, `strict`, and `disabled`. Basically, `enabled` mode enables non-approved FIPS whereas `strict` enables approved FIPS (more strict about algorithms, keys, etc).
* Users only need to copy BCFIPS jars to `providers` dir and choose one of the FIPS modes. I did not find a way to automatically enable FIPS based on the availability of BCFIPS providers. There is a class loader constraint.
* Users can easily enable/disable FIPS without removing BCFIPS jars from the `providers` dir.

Example:

```
kc.sh start --security-fips-mode strict
```

To dump the security providers installed:

```
kc.sh start --log-level org.keycloak.common.crypto.CryptoIntegration:trace --security-fips-mode enabled
```

Please, let me know if the configuration model and UX make sense.

<!---
Please read https://github.com/keycloak/keycloak/blob/main/CONTRIBUTING.md and follow these guidelines when contributing to Keycloak
-->

pedroigor committed:
Make it possible to switch between BC and BC-FIPS libraries

Closes #12424

mposolda reviewed (commented):
@pedroigor Thanks Pedro! The PR looks great to me. I have few minor comments/questions:

- The option `security-fips-mode` requires BCFIPS on the classpath, is it correct? Should this be directly added to the help text of the option? Or will the help text be too long after this (Hence better to just stick with this info in the docs?)

- The docs is missing, but I guess you are aware for that? For the testing, I am planning to start some automated tests for FIPS in integration testsuite soon, so I think you don't need to care about tests in this PR though (unless you plan some unit tests in quarkus module or something)?

- Should we auto-enable FIPS when BCFIPS are on classpath? However I am not sure whether to use `enabled` or `strict` mode. So maybe rather leaving as is, so the admin needs to actively decide between those 2 modes?

pedroigor commented:
> @pedroigor Thanks Pedro! The PR looks great to me. I have few minor comments/questions:
> 
> * The option `security-fips-mode` requires BCFIPS on the classpath, is it correct? Should this be directly added to the help text of the option? Or will the help text be too long after this (Hence better to just stick with this info in the docs?)

IMO, docs. Otherwise, too long help message.

> * The docs is missing, but I guess you are aware for that? For the testing, I am planning to start some automated tests for FIPS in integration testsuite soon, so I think you don't need to care about tests in this PR though (unless you plan some unit tests in quarkus module or something)?

Yes, both docs and tests are missing hence the label applied to this PR.

I think we do need to care about tests and docs, and I should be writing one.

> * Should we auto-enable FIPS when BCFIPS are on classpath? However I am not sure whether to use `enabled` or `strict` mode. So maybe rather leaving as is, so the admin needs to actively decide between those 2 modes?

As mentioned in the description, there are some class loading constraints that block us from doing this.

pedroigor commented:
@mposolda I've added a test to the distribution.

W.r.t. to docs, I think we can skip it for now as this is a WIP. So we can come up with a proper guide later when have more things to add there.

I also changed the option to `fips-mode` instead of `security-fips-mode`. Not sure if the best but it is less verbose and we don't quite need the `security` prefix ... However, I think it will impact the operator config. @vmuzikar.

mposolda reviewed (approved):
@pedroigor Thanks Pedro! +1 to adding the docs as a follow-up . I am approving and would be happy to have this available ASAP in keycloak main, so I can use it :)
